src/raftstore/store/snap.rs

```diff
@@ -1302,15 +1302,19 @@ impl SnapManager {
         self.max_total_size
     }
 
-    pub fn register(&self, key: SnapKey, entry: SnapEntry) {
+    // For `SnapEntry::Sending` we need to support register multi times because
+    // Raft can send same snapshot to many peers (e.g. a follower and several learners).
+    // But for other kinds, Only support registering at most one time for each of them.
+    // TODO: make it work as a state machine, and change `Vec<SnapEntry>` to `SnapEntry`.
```

Comment:
Why not do it now?

Reply:
It's better to do it after we change the snapshot generating API. Suppose trying to register `Generating` again when the snapshot is already `Sending`. It's not necessary to let the re-gister return fail, but otherwise it will also be more complex when deregister it. So I think we can do it after changing the snapshot generating API. After that we can avoid manage `Register/Deregister` explicitly.

Reply:
In your changes, it returns failure already.